packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,513 @@
+  /// The identifier of the widget that is currently selected.
+  ///
+  /// This must be one of the keys in the [Map] of [children].
+  /// If this attribute is null, no widget will be initially selected.
+  final T groupValue;
+
+  /// The callback that is called when a new option is tapped.
+  ///
+  /// This attribute must not be null.
+  ///
+  /// The segmented control passes the newly selected widget's associated key
+  /// to the callback but does not actually change state until the parent
+  /// widget rebuilds the segmented control with the new [groupValue].
+  ///
+  /// The callback provided to [onValueChanged] should update the state of
+  /// the parent [StatefulWidget] using the [State.setState] method, so that
+  /// the parent gets rebuilt; for example:
+  ///
+  /// ## Sample code
+  ///
+  /// ```dart
+  /// class SegmentedControlExample extends StatefulWidget {
+  ///   @override
+  ///   State createState() => new SegmentedControlExampleState();
+  /// }
+  ///
+  /// class SegmentedControlExampleState extends State<SegmentedControlExample> {
+  ///   final Map<int, Widget> children = const {
+  ///     0: const Text('Child 1'),
+  ///     1: const Text('Child 2'),
+  ///   };
+  ///
+  ///   int currentValue;
+  ///
+  ///   @override
+  ///   Widget build(BuildContext context) {
+  ///     return new Container(
+  ///       child: new SegmentedControl<int>(
+  ///         children: children,
+  ///         onValueChanged: (int newValue) {
+  ///           setState(() {
+  ///             currentValue = newValue;
+  ///           });
+  ///         },
+  ///         groupValue: currentValue,
+  ///       ),
+  ///     );
+  ///   }
+  /// }
+  /// ```
+  final ValueChanged<T> onValueChanged;
+
+  @override
+  _SegmentedControlState<T> createState() => _SegmentedControlState<T>();
+}
+
+class _SegmentedControlState<T> extends State<SegmentedControl<T>> {
+  int _selectedIndex;
+  int _pressedIndex;
+  Map<T, int> keysToIndices = <T, int>{};
+  @override
+  void initState() {
+    super.initState();
+    for (int index = 0; index < widget.children.keys.length; index += 1) {
+      keysToIndices[widget.children.keys.elementAt(index)] = index;
+    }
+    if (widget.children.keys.contains(widget.groupValue)) {
+      _selectedIndex = keysToIndices[widget.groupValue];
+    }
+  }
+
+  void _onTapDown(T currentKey) {
+    setState(() {
+      _pressedIndex = keysToIndices[currentKey];
+    });
+  }
+
+  void _onTapUp(TapUpDetails event) {
+    setState(() {
+      _pressedIndex = null;
+    });
+  }
+
+  void _onTapCancel() {
+    setState(() {
+      _pressedIndex = null;
+    });
+  }
+
+  void _onTap(T currentKey) {
+    setState(() {
+      if (_selectedIndex != keysToIndices[currentKey]) {
+        widget.onValueChanged(currentKey);
+      }
+    });
+  }
+
+  @override
+  void didUpdateWidget(SegmentedControl<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.children.keys.contains(widget.groupValue)) {
```

Comment:
Same here as for initState: just assert and check for null.

And also, even if you did need to do this, you can use "widget.children.contains" instead of creating the list of keys, and the bonus is that it'll be faster to search that way because it's searching a hash map instead of a list. (O(1) instead of O(n)).